MC-32488 MySQL 8 update (#7283)
magento/devdocs: src/guides/v2.4/install-gde/prereq/mysql.md

| user\_id | int unsigned | NO | PRI | NULL | auto\_increment |
| firstname | varchar\(32\) | YES | | NULL | |
| lastname | varchar\(32\) | YES | | NULL | |
| email | varchar\(128\) | YES | | NULL | |
| username | varchar\(40\) | YES | UNI | NULL | |
| password | varchar\(255\) | NO | | NULL | |
| created | timestamp | NO | | CURRENT\_TIMESTAMP | DEFAULT\_GENERATED |
| modified | timestamp | NO | | CURRENT\_TIMESTAMP | DEFAULT\_GENERATED on update CURRENT\_TIMESTAMP |
| logdate | timestamp | YES | | NULL | |
| lognum | smallint unsigned | NO | | 0 | |

With the exception of *TINYINT(1)*, all integer padding (TINYINT > 1, SMALLINT, MEDIUMINT, INT, BIGINT) should be removed from `db_schema.xml`.

For more information, see [https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-19.html#mysqld-8-0-19-feature](
https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-19.html#mysqld-8-0-19-feature).

### Default ORDER BY behavior

Before 8.0, entries were sorted by the foreign key. Default sort order depends on the engine that is used.
Always specify a sort order if your code depends on a specific sort.

### Deprecated ASC and DESC qualifiers for GROUP BY

> As of MySQL 8.0.13, the deprecated `ASC` or `DESC` qualifiers for `GROUP BY` clauses have been removed. Queries that previously relied on GROUP BY sorting may produce results that differ from previous MySQL versions. To produce a given sort order, provide an ORDER BY clause.

Reviewer: ```suggestion
As of MySQL 8.0.13, the deprecated `ASC` or `DESC` qualifiers for `GROUP BY` clauses have been removed. Queries that previously relied on `GROUP BY` sorting may produce results that differ from previous MySQL versions. To produce a given sort order, provide an `ORDER BY` clause.
```

## Magento and MySQL 8

There have been some changes to Magento to properly support MySQL 8.

### Query and Insert Behavior

Magento disabled the regular validation behavior by setting SET SQL_MODE='' in `/lib/internal/Magento/Framework/DB/Adapter/Pdo/Mysql.php:424.`. With validation disabled, it is possible that MySQL will truncate data. In MySQL, the Query behavior has changed: `Select * on my_table where IP='127.0.0.1'` will no longer return any results because the IP address is now properly seen as a string, rather than an integer.

## Upgrading from MySQL 5.7 to MySQL 8

> In order to properly update your MySQL from version 5.7 to version 8, the process must be done in the following order:

Reviewer: ```suggestion
To properly update MySQL from version 5.7 to version 8, you must follow these steps in order:
```

1. Upgrade Magento to 2.4.0.

>    Test everything makes sure your system work as expected.

Reviewer: ```suggestion
   Test everything and make sure your system works as expected.
```

1. Enable maintenance mode.
1. Make a backup from the database.
1. Update MySQL to version 8.

> 1. Import the dump into MySQL.

Reviewer: Is the "dump" the same thing as the "backup" mentioned in step 3? If so, use the same terminology.

> 1. Switch engine clean cache.

Reviewer: Is this two steps?

> 1. Disable maintenance mode.

Reviewer: Show the command.

Always specify a sort order if your code depends on a specific sort.

### Deprecated ASC and DESC qualifiers for GROUP BY

As of MySQL 8.0.13, the deprecated `ASC` or `DESC` qualifiers for `GROUP BY` clauses have been removed. Queries that previously relied on `GROUP BY` sorting may produce results that differ from previous MySQL versions. To produce a given sort order, provide an `ORDER BY` clause.

## Magento and MySQL 8

There have been some changes to Magento to properly support MySQL 8.

### Query and Insert Behavior

Magento disabled the regular validation behavior by setting SET SQL_MODE='' in `/lib/internal/Magento/Framework/DB/Adapter/Pdo/Mysql.php:424.`. With validation disabled, it is possible that MySQL will truncate data. In MySQL, the Query behavior has changed: `Select * on my_table where IP='127.0.0.1'` will no longer return any results because the IP address is now properly seen as a string, rather than an integer.

## Upgrading from MySQL 5.7 to MySQL 8

To properly update MySQL from version 5.7 to version 8, you must follow these steps in order:

1. Upgrade Magento to 2.4.0.
   Test everything and make sure your system works as expected.

> 1. Enable maintenance mode: `bin/magento maintenance:enable`.

Reviewer: Commands should be added below the step, indented 3 spaces, with code fences.

1. Make a data [backup]({{page.baseurl }}/install-gde/install/cli/install-cli-backup).html#instgde-cli-uninst-back) from the database.
1. Update MySQL to version 8.
1. Import the backed-up data into MySQL.

> 1. Set the database engine in the Magento admin to MySQL 8.

Reviewer: Is there a `bin/magento` command for that?

Author: I can't find one.

Reviewer: See https://devdocs.magento.com/guides/v2.3/reference/cli/magento.html#setupinstall

> 1. Clean the [cache](https://docs.magento.com/user-guide/system/cache-management.html).

Reviewer: I think it makes more sense to add the CLI commands for clearing the cache rather than linking to the user guide, which shows you how to do it in the UI.

> 1. Disable maintenance mode: `bin/magento maintenance:disable`.

Reviewer: Commands should be added below the step, indented 3 spaces, with code fences.
